### PR TITLE
Add test c10d ucc tests

### PR DESCRIPTION
**Pull request by Fuzzkatt**
Creates the equivalent c10d test for ucc for https://github.com/pytorch/pytorch/blob/master/test/distributed/test_c10d_gloo.py and https://github.com/pytorch/pytorch/blob/master/test/distributed/test_c10d_nccl.py. Uses test_c10d_gloo.py as the reference and adds all the common ops. More detailed comparison of available ops here: https://docs.google.com/document/d/1yPsa_X9EiEiqo-j2Yn7ierhccBtEjwoqC-B7-amI0MI/edit?usp=sharing

Also removes extra line for ProcessGroupUCC.cpp barrier blocking wait that got duplicated from merging https://github.com/pytorch/pytorch/pull/85047.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu

### CODE REVIEW

**Comment by Fuzzkatt** [comment]
Ready for review

**Comment by kwen2501** [comment]
Cc @vtlam @kingchc . Can you please help take a look? Thanks!

**Review by zasdfgbnm — COMMENTED** [review]
Finished reading. Lots of my comments are about adding new tests, considering the size of this PR, I these comments are better resolved in followup PRs. For this PR, I also have a few comments, please fix them here.

File: test/distributed/test_c10d_ucc.py [hunk 1]
```diff
@@ -0,0 +1,1115 @@
...
+    def _create_process_group_ucc(self):
+        store = c10d.FileStore(self.file_name, self.world_size)
+        return c10d.ProcessGroupUCC(store, self.rank, self.world_size)
+
+    def setUp(self):
+        super(ProcessGroupUCCTest, self).setUp()
+        self._spawn_processes()
+
+    def tearDown(self):
+        super(ProcessGroupUCCTest, self).tearDown()
```
[the remark this answers is not in the document]
Comment (Fuzzkatt) [reply test/distributed/test_c10d_ucc.py]: fixed

File: test/distributed/test_c10d_ucc.py [hunk 2]
```diff
@@ -0,0 +1,1115 @@
...
+            fut.wait()
+            return fut.value()
+
+        # Every rank is root once
+        for i in range(self.world_size):
+            # Run with 1 input tensor
+            x = fn(torch.tensor([self.rank]))
+            output = broadcast([x], i, 0)
+            # TODO(#38095): Replace assertEqualIgnoreType. See issue #38095
+            self.assertEqualIgnoreType(torch.tensor([i]), output[0])
```
[the remark this answers is not in the document]
Comment (Fuzzkatt) [reply test/distributed/test_c10d_ucc.py]: fixed

File: test/distributed/test_c10d_ucc.py [hunk 3]
```diff
@@ -0,0 +1,1115 @@
...
+                opts = c10d.ReduceOptions()
+                opts.reduceOp = op
+                opts.rootRank = root
+                tmp = fn(input)
+                fut = pg.reduce([tmp], opts).get_future()
+                fut.wait()
+                result = fut.value()
+                if root == self.rank:
+                    # TODO(#38095): Replace assertEqualIgnoreType. See issue #38095
+                    self.assertEqualIgnoreType(output, result[0])
```
[the remark this answers is not in the document]
Comment (Fuzzkatt) [reply test/distributed/test_c10d_ucc.py]: fixed

File: test/distributed/test_c10d_ucc.py [hunk 4]
```diff
@@ -0,0 +1,1115 @@
...
+        tests = simple_reduce_tests(self.rank, self.world_size)
+        for (op, input, expected) in tests:
+            opts = c10d.AllreduceOptions()
+            opts.reduceOp = op
+            tensor = fn(input)
+            fut = pg.allreduce([tensor], opts).get_future()
+            fut.wait()
+            result = fut.value()
+            # TODO(#38095): Replace assertEqualIgnoreType. See issue #38095
+            self.assertEqualIgnoreType(expected, result[0])
```
[the remark this answers is not in the document]
Comment (Fuzzkatt) [reply test/distributed/test_c10d_ucc.py]: fixed

File: test/distributed/test_c10d_ucc.py [hunk 5]
```diff
@@ -0,0 +1,1115 @@
...
+        )
+
+    @skip_if_lt_x_gpu(2)
+    def test_broadcast_work_wait_gpu(self):
+        self._test_broadcast_work_wait(
+            torch.ones(2, 2, device=self.rank) * self.rank
+        )
+
+    @skip_if_lt_x_gpu(2)
+    def test_nested_comm_tensor_wrapping(self):
```
[the remark this answers is not in the document]
Comment (Fuzzkatt) [reply test/distributed/test_c10d_ucc.py]: fixed, will add

File: test/distributed/test_c10d_ucc.py [hunk 6]
```diff
@@ -0,0 +1,1115 @@
...
+        super(CommTest, self).setUp()
+        self._spawn_processes()
+
+    def tearDown(self):
+        super(CommTest, self).tearDown()
+        try:
+            os.remove(self.file_name)
+        except OSError:
+            pass
+
```
[the remark this answers is not in the document]
Comment (Fuzzkatt) [reply test/distributed/test_c10d_ucc.py:944]: Yes, I'll remove the skipped tests for now

File: test/distributed/test_c10d_ucc.py [hunk 7]
```diff
@@ -0,0 +1,1115 @@
...
+                c10d.ReduceOp.BXOR,
+                torch.tensor([vin], dtype=torch.int32),
+                torch.tensor([vout], dtype=torch.int32),
+            ),
+        )
+
+    return tests
+
+
+def simple_coalesced_reduce_tests(rank, world_size):
```
[the remark this answers is not in the document]
Comment (Fuzzkatt) [reply test/distributed/test_c10d_ucc.py]: no, removed

File: test/distributed/test_c10d_ucc.py [hunk 8]
```diff
@@ -0,0 +1,1115 @@
...
+        ),
+        (
+            c10d.ReduceOp.MAX,
+            [torch.tensor([rank + x]) for x in [1.0, 2.0]],
+            [torch.tensor([world_size]), torch.tensor([world_size + 1.0])],
+        ),
+    ]
+
+
+def simple_multi_input_reduce_tests(rank, world_size):
```
[the remark this answers is not in the document]
Comment (Fuzzkatt) [reply test/distributed/test_c10d_ucc.py]: no, removed

File: test/distributed/test_c10d_ucc.py [hunk 9]
```diff
@@ -0,0 +1,1115 @@
...
+        self._test_default_store_timeout("ucc")
+
+
+class ProcessGroupUCCTest(MultiProcessTestCase):
+    def _create_process_group_ucc(self):
+        store = c10d.FileStore(self.file_name, self.world_size)
+        return c10d.ProcessGroupUCC(store, self.rank, self.world_size)
+
+    def setUp(self):
+        super(ProcessGroupUCCTest, self).setUp()
```
[the remark this answers is not in the document]
Comment (Fuzzkatt) [reply test/distributed/test_c10d_ucc.py]: fixed

File: test/distributed/test_c10d_ucc.py [hunk 10]
```diff
@@ -0,0 +1,1115 @@
...
+            backend="ucc",
+            rank=self.rank,
+            world_size=self.world_size,
+            store=store,
+        )
+        return dist.distributed_c10d._get_default_group()
+
+    @skip_if_lt_x_gpu(2)
+    def test_allreduce_work_wait_gpu(self):
+        self._test_allgather_work_wait(
```
[the remark this answers is not in the document]
Comment (Fuzzkatt) [reply test/distributed/test_c10d_ucc.py]: Good catch, fixed.

**Review by zasdfgbnm — APPROVED** [review]
LGTM

File: test/distributed/test_c10d_ucc.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (kit1980) [reply test/distributed/test_c10d_ucc.py]: I see there are more of these, line 167, for example.
All existing #38095-related issues were fixed some time ago and the issue was closed, please don't re-introduce.

**Review by kit1980 — CHANGES_REQUESTED** [review]
All existing https://github.com/pytorch/pytorch/issues/38095-related issues were fixed some time ago and the issue was closed, please don't re-introduce.

**Comment by Fuzzkatt** [comment]
> All existing #38095 issues were fixed some time ago and the issue was closed, please don't re-introduce.

Sorry about that, the comments referencing #38095 were outdated (this PR is a couple months old) and I removed them. There are no actual issues relating to #38905.

**Comment by zasdfgbnm** [comment]
@kit1980 Do you still have something to request change or are you OK with the current status?

**Review by kit1980 — APPROVED** [review]

**Review by kwen2501 — APPROVED** [review]

**Review by malfet — APPROVED** [review]

**Pull request closed without merging**